File: server/apps/api/src/animedownloader_api/routes/release_profiles.py
from typing import Annotated
from typing import Annotated
from uuid import UUID

from animedownloader_releases import (
    InvalidReleaseParserProfileError,
    ParserField,
    ParserRuleSpec,
    ParserTransform,
    ReleaseGroupNotFoundError,
    ReleaseParserObservationNotFoundError,
    ReleaseParserProfile,
    ReleaseParserProfileActivationError,
    ReleaseParserProfileNotFoundError,
    ReleaseParserSampleNotFoundError,
    ReleaseProfileService,
)
from fastapi import APIRouter, Depends, HTTPException, status

from animedownloader_api.dependencies import get_release_profile_service
from animedownloader_api.release_profile_schemas import (
    ParserComparisonResponse,
    ParserDifferenceResponse,
    ParserHealthResponse,
    ParserSampleResultResponse,
    ParserValidationResponse,
    ReleaseGroupSummaryResponse,
    ReleaseParserObservationResponse,
    ReleaseParserProfileResponse,
    ReleaseParserProfileUpdate,
    ReleaseParserRuleResponse,
    ReleaseParserSampleCreate,
    ReleaseParserSampleResponse,
)
from animedownloader_api.schemas import ParsedReleaseResponse

router = APIRouter(tags=["release-profiles"])
ServiceDependency = Annotated[
    ReleaseProfileService,
    Depends(get_release_profile_service),
]


@router.get("/api/release-groups", response_model=list[ReleaseGroupSummaryResponse])
async def list_release_groups(
    service: ServiceDependency,
) -> list[ReleaseGroupSummaryResponse]:
    groups = await service.list_groups()
    response: list[ReleaseGroupSummaryResponse] = []
    for group in groups:
        active_version = next(
            (
                profile.version
                for profile in reversed(group.parser_profiles)
                if profile.status == "active"
            ),
            None,
        )
        draft_version = next(
            (
                profile.version
                for profile in reversed(group.parser_profiles)
                if profile.status == "draft"
            ),
            None,
        )
        response.append(
            ReleaseGroupSummaryResponse(
                id=group.id,
                name=group.name,
                slug=group.slug,
                enabled=group.enabled,
                active_parser_profile_version=active_version,
                draft_parser_profile_version=draft_version,
            ),
        )
    return response


@router.get(
    "/api/release-groups/{group_id}/parser-profiles",
    response_model=list[ReleaseParserProfileResponse],
)
async def list_parser_profiles(
    group_id: UUID,
    service: ServiceDependency,
) -> list[ReleaseParserProfileResponse]:
    try:
        profiles = await service.list_profiles(group_id)
    except ReleaseGroupNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return [_profile_response(profile) for profile in profiles]


@router.post(
    "/api/release-groups/{group_id}/parser-profiles/drafts",
    response_model=ReleaseParserProfileResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_parser_draft(
    group_id: UUID,
    service: ServiceDependency,
) -> ReleaseParserProfileResponse:
    try:
        profile = await service.create_draft(group_id)
    except ReleaseGroupNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return _profile_response(profile)


@router.get(
    "/api/release-parser-profiles/{profile_id}",
    response_model=ReleaseParserProfileResponse,
)
async def get_parser_profile(
    profile_id: UUID,
    service: ServiceDependency,
) -> ReleaseParserProfileResponse:
    try:
        profile = await service.get_profile(profile_id)
    except ReleaseParserProfileNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return _profile_response(profile)


@router.patch(
    "/api/release-parser-profiles/{profile_id}",
    response_model=ReleaseParserProfileResponse,
)
async def update_parser_profile(
    profile_id: UUID,
    payload: ReleaseParserProfileUpdate,
    service: ServiceDependency,
) -> ReleaseParserProfileResponse:
    rules = tuple(
        ParserRuleSpec(
            field=item.field,
            pattern=item.pattern,
            priority=item.priority,
            required=item.required,
            flags=item.flags,
            transform=item.transform,
        )
        for item in payload.rules
    )
    try:
        profile = await service.update_rules(profile_id, rules)
    except ReleaseParserProfileNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except InvalidReleaseParserProfileError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    return _profile_response(profile)


@router.get(
    "/api/release-groups/{group_id}/parser-samples",
    response_model=list[ReleaseParserSampleResponse],
)
async def list_parser_samples(
    group_id: UUID,
    service: ServiceDependency,
) -> list[ReleaseParserSampleResponse]:
    try:
        samples = await service.list_samples(group_id)
    except ReleaseGroupNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return [ReleaseParserSampleResponse.model_validate(sample) for sample in samples]


@router.post(
    "/api/release-groups/{group_id}/parser-samples",
    response_model=ReleaseParserSampleResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_parser_sample(
    group_id: UUID,
    payload: ReleaseParserSampleCreate,
    service: ServiceDependency,
) -> ReleaseParserSampleResponse:
    try:
        sample = await service.add_sample(
            group_id,
            title=payload.title,
            source=payload.source,
        )
    except ReleaseGroupNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    return ReleaseParserSampleResponse.model_validate(sample)


@router.delete(
    "/api/release-parser-samples/{sample_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def delete_parser_sample(
    sample_id: UUID,
    service: ServiceDependency,
) -> None:
    try:
        await service.delete_sample(sample_id)
    except ReleaseParserSampleNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.post(
    "/api/release-parser-profiles/{profile_id}/validate",
    response_model=ParserValidationResponse,
)
async def validate_parser_profile_endpoint(
    profile_id: UUID,
    service: ServiceDependency,
) -> ParserValidationResponse:
    try:
        result = await service.validate_profile(profile_id)
    except ReleaseParserProfileNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return ParserValidationResponse(
        valid=result.valid,
        sample_count=result.sample_count,
        minimum_samples=result.minimum_samples,
        errors=list(result.errors),
        results=[
            ParserSampleResultResponse(
                sample_id=item.sample_id,
                title=item.title,
                parsed=ParsedReleaseResponse.from_parsed(item.parsed)
                if item.parsed is not None
                else None,
                error=item.error,
            )
            for item in result.results
        ],
    )


@router.get(
    "/api/release-parser-profiles/{profile_id}/comparison",
    response_model=ParserComparisonResponse,
)
async def compare_parser_profile(
    profile_id: UUID,
    service: ServiceDependency,
) -> ParserComparisonResponse:
    try:
        result = await service.compare_with_active(profile_id)
    except ReleaseParserProfileNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return ParserComparisonResponse(
        profile_id=result.profile_id,
        draft_version=result.draft_version,
        active_version=result.active_version,
        differences=[
            ParserDifferenceResponse(
                sample_id=item.sample_id,
                title=item.title,
                fields={
                    field: [before, after]
                    for field, (before, after) in item.fields.items()
                },
            )
            for item in result.differences
        ],
    )


@router.get(
    "/api/release-parser-profiles/{profile_id}/health",
    response_model=ParserHealthResponse,
)
async def parser_profile_health(
    profile_id: UUID,
    service: ServiceDependency,
) -> ParserHealthResponse:
    try:
        result = await service.health(profile_id)
    except ReleaseParserProfileNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return ParserHealthResponse(
        profile_id=result.profile_id,
        total_count=result.total_count,
        parsed_count=result.parsed_count,
        ambiguous_count=result.ambiguous_count,
        unparsed_count=result.unparsed_count,
        unsupported_count=result.unsupported_count,
        failure_rate=result.failure_rate,
        drift_signal=result.drift_signal,
        drift_reason=result.drift_reason,
        recent_failures=[
            ReleaseParserObservationResponse.model_validate(item)
            for item in result.recent_failures
        ],
    )


@router.post(
    "/api/release-parser-profiles/{profile_id}/activate",
    response_model=ReleaseParserProfileResponse,
)
async def activate_parser_profile(
    profile_id: UUID,
    service: ServiceDependency,
) -> ReleaseParserProfileResponse:
    try:
        profile = await service.activate(profile_id)
    except ReleaseParserProfileNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ReleaseParserProfileActivationError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    return _profile_response(profile)


@router.post(
    "/api/release-parser-observations/{observation_id}/draft",
    response_model=ReleaseParserProfileResponse,
)
async def create_draft_from_observation(
    observation_id: UUID,
    service: ServiceDependency,
) -> ReleaseParserProfileResponse:
    try:
        profile = await service.create_draft_from_observation(observation_id)
    except ReleaseParserObservationNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return _profile_response(profile)


def _profile_response(
    profile: ReleaseParserProfile,
) -> ReleaseParserProfileResponse:
    return ReleaseParserProfileResponse(
        id=profile.id,
        release_group_id=profile.release_group_id,
        release_group_name=profile.release_group.name,
        version=profile.version,
        status=profile.status,
        created_at=profile.created_at,
        activated_at=profile.activated_at,
        rules=[
            ReleaseParserRuleResponse(
                id=rule.id,
                field=rule.field,
                pattern=rule.pattern,
                priority=rule.priority,
                required=rule.required,
                flags=rule.flags,
                transform=rule.transform,
            )
            for rule in profile.rules
        ],
    )

File: web/apps/web/src/entities/release/api/releaseProfiles.ts
import { z } from "zod";

export const parserFields = [
  "release_group",
  "series_title",
  "episode_number",
  "episode_title",
  "season_number",
  "resolution",
  "source",
  "video_codec",
  "audio_codec",
  "bit_depth",
] as const;
import {
  deleteJson,
  getJson,
  patchJson,
  postJson,
} from "../../../shared/api/client";

const releaseGroupSchema = z.object({
  id: z.uuid(),
  name: z.string(),
  slug: z.string(),
  enabled: z.boolean(),
  active_parser_profile_version: z.number().int().positive().nullable(),
  draft_parser_profile_version: z.number().int().positive().nullable(),
});

const parserRuleSchema = z.object({
  id: z.uuid(),
  field: z.enum(parserFields),
  pattern: z.string(),
  priority: z.number().int(),
  required: z.boolean(),
  flags: z.string(),
  transform: z.enum([
    "identity",
    "strip",
    "normalize_spaces",
    "to_int",
    "lower",
    "upper",
  ]),
});

const parserProfileSchema = z.object({
  id: z.uuid(),
  release_group_id: z.uuid(),
  release_group_name: z.string(),
  version: z.number().int().positive(),
  status: z.enum(["draft", "active", "retired"]),
  created_at: z.coerce.date(),
  activated_at: z.coerce.date().nullable(),
  rules: z.array(parserRuleSchema),
});

const sampleSchema = z.object({
  id: z.uuid(),
  release_group_id: z.uuid(),
  source: z.string(),
  title: z.string(),
  created_at: z.coerce.date(),
  updated_at: z.coerce.date(),
});

const observationSchema = z.object({
  id: z.uuid(),
  release_group_id: z.uuid(),
  profile_id: z.uuid(),
  source: z.string(),
  title: z.string(),
  status: z.string(),
  failed_required_fields: z.array(z.string()),
  observed_at: z.coerce.date(),
});

const parsedReleaseSchema = z.object({
  provider_source: z.string(),
  source_id: z.string(),
  original_title: z.string(),
  normalized_title: z.string(),
  release_group: z.string().nullable(),
  series_title: z.string().nullable(),
  episode_number: z.number().int().positive().nullable(),
  episode_title: z.string().nullable(),
  season_number: z.number().int().positive().nullable(),
  resolution: z.string().nullable(),
  source: z.string().nullable(),
  video_codec: z.string().nullable(),
  audio_codec: z.string().nullable(),
  bit_depth: z.number().int().positive().nullable(),
  status: z.enum(["parsed", "ambiguous", "unparsed", "unsupported"]),
  warnings: z.array(z.string()),
  failed_required_fields: z.array(z.string()),
  parser_profile_version: z.number().int().positive().nullable(),
});

const validationSchema = z.object({
  valid: z.boolean(),
  sample_count: z.number().int().nonnegative(),
  minimum_samples: z.number().int().positive(),
  errors: z.array(z.string()),
  results: z.array(
    z.object({
      sample_id: z.uuid(),
      title: z.string(),
      parsed: parsedReleaseSchema.nullable(),
      error: z.string().nullable(),
    }),
  ),
});

const comparisonSchema = z.object({
  profile_id: z.uuid(),
  draft_version: z.number().int().positive(),
  active_version: z.number().int().positive().nullable(),
  differences: z.array(
    z.object({
      sample_id: z.uuid(),
      title: z.string(),
      fields: z.record(z.string(), z.array(z.unknown())),
    }),
  ),
});

const healthSchema = z.object({
  profile_id: z.uuid(),
  total_count: z.number().int().nonnegative(),
  parsed_count: z.number().int().nonnegative(),
  ambiguous_count: z.number().int().nonnegative(),
  unparsed_count: z.number().int().nonnegative(),
  unsupported_count: z.number().int().nonnegative(),
  failure_rate: z.number().min(0).max(1),
  drift_signal: z.boolean(),
  drift_reason: z.string().nullable(),
  recent_failures: z.array(observationSchema),
});

export type ReleaseGroupSummary = z.infer<typeof releaseGroupSchema>;
export type ParserRule = z.infer<typeof parserRuleSchema>;
export type ParserProfile = z.infer<typeof parserProfileSchema>;
export type ParserSample = z.infer<typeof sampleSchema>;
export type ParserObservation = z.infer<typeof observationSchema>;
export type ParserValidation = z.infer<typeof validationSchema>;
export type ParserComparison = z.infer<typeof comparisonSchema>;
export type ParserHealth = z.infer<typeof healthSchema>;

export type ParserRuleInput = Omit<ParserRule, "id">;

export class ReleaseProfileResponseError extends Error {
  constructor(readonly issues: z.core.$ZodIssue[]) {
    super(
      "Invalid release profile API response: " +
        issues
          .map(
            (issue) =>
              (issue.path.join(".") || "<root>") + ": " + issue.message,
          )
          .join("; "),
    );
    this.name = "ReleaseProfileResponseError";
  }
}

export async function createReleaseGroup(
  input: { name: string; slug?: string },
  signal?: AbortSignal,
): Promise<ReleaseGroupSummary> {
  return parseSchema(
    await postJson("/api/release-groups", input, { signal }),
    releaseGroupSchema,
  );
}

export async function getReleaseGroups(
  signal?: AbortSignal,
): Promise<ReleaseGroupSummary[]> {
  return parseList(await getJson("/api/release-groups", { signal }), releaseGroupSchema);
}

export async function getParserProfiles(
  groupId: string,
  signal?: AbortSignal,
): Promise<ParserProfile[]> {
  return parseList(
    await getJson("/api/release-groups/" + groupId + "/parser-profiles", { signal }),
    parserProfileSchema,
  );
}

export async function createParserDraft(
  groupId: string,
  signal?: AbortSignal,
): Promise<ParserProfile> {
  return parseParserProfile(
    await postJson("/api/release-groups/" + groupId + "/parser-profiles/drafts", {}, { signal }),
  );
}

export async function updateParserProfile(
  profileId: string,
  rules: ParserRuleInput[],
  signal?: AbortSignal,
): Promise<ParserProfile> {
  return parseParserProfile(
    await patchJson(
      "/api/release-parser-profiles/" + profileId,
      { rules },
      { signal },
    ),
  );
}

export async function getParserSamples(
  groupId: string,
  signal?: AbortSignal,
): Promise<ParserSample[]> {
  return parseList(
    await getJson("/api/release-groups/" + groupId + "/parser-samples", { signal }),
    sampleSchema,
  );
}

export async function createParserSample(
  groupId: string,
  input: { title: string; source: string },
  signal?: AbortSignal,
): Promise<ParserSample> {
  return parseSchema(
    await postJson("/api/release-groups/" + groupId + "/parser-samples", input, { signal }),
    sampleSchema,
  );
}

export async function deleteParserSample(
  sampleId: string,
  signal?: AbortSignal,
): Promise<void> {
  await deleteJson("/api/release-parser-samples/" + sampleId, { signal });
}

export async function validateParserProfile(
  profileId: string,
  signal?: AbortSignal,
): Promise<ParserValidation> {
  return parseSchema(
    await postJson("/api/release-parser-profiles/" + profileId + "/validate", {}, { signal }),
    validationSchema,
  );
}

export async function compareParserProfile(
  profileId: string,
  signal?: AbortSignal,
): Promise<ParserComparison> {
  return parseSchema(
    await getJson("/api/release-parser-profiles/" + profileId + "/comparison", { signal }),
    comparisonSchema,
  );
}

export async function getParserHealth(
  profileId: string,
  signal?: AbortSignal,
): Promise<ParserHealth> {
  return parseSchema(
    await getJson("/api/release-parser-profiles/" + profileId + "/health", { signal }),
    healthSchema,
  );
}

export async function activateParserProfile(
  profileId: string,
  signal?: AbortSignal,
): Promise<ParserProfile> {
  return parseParserProfile(
    await postJson("/api/release-parser-profiles/" + profileId + "/activate", {}, { signal }),
  );
}

export async function createDraftFromObservation(
  observationId: string,
  signal?: AbortSignal,
): Promise<ParserProfile> {
  return parseParserProfile(
    await postJson(
      "/api/release-parser-observations/" + observationId + "/draft",
      {},
      { signal },
    ),
  );
}

function parseParserProfile(payload: unknown): ParserProfile {
  return parseSchema(payload, parserProfileSchema);
}

function parseSchema<T extends z.ZodType>(
  payload: unknown,
  schema: T,
): z.infer<T> {
  const result = schema.safeParse(payload);
  if (!result.success) {
    throw new ReleaseProfileResponseError(result.error.issues);
  }
  return result.data;
}

function parseList<T extends z.ZodType>(
  payload: unknown,
  schema: T,
): z.infer<T>[] {
  return parseSchema(payload, z.array(schema));
}
